File: Day_3_Rucksack_Reorganization/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"unicode/utf8"
)

// get line-by-line file scanner
func GetFileScanner(filePath string) (*bufio.Scanner, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file %s, %s", filePath, err.Error())
	}

	scanner := bufio.NewScanner(f)

	return scanner, nil
}

// get priority of item based on ASCII value
func GetItemPriority(item string) int {
	itemASCII := int([]rune(item)[0])
	aASCII := int('a')
	AASCII := int('A')
	zASCII := int('z')

	// when item is lowercase, priority is given by how many steps ahead of 'a' the item is
	itemPriority := itemASCII - aASCII + 1
	// when item is uppercase
	if itemASCII < aASCII {
		// priority is given by how many steps ahead of 'A' the item is
		// plus the number of letters in the alphabet (i.e. 'z' - 'a' + 1)
		itemPriority = itemASCII - AASCII + 1 + zASCII - aASCII + 1
	}

	return itemPriority
}

// identify duplicate item between two given strings
func GetDuplicateItems(items1 string, items2 string) string {
	// hash map of elements in items1
	items1Map := make(map[string]bool)
	// hash map of duplicate elements
	duplicateItemsMap := make(map[string]bool)

	// loop over items1 and store all items
	for i := 0; i < len(items1); i++ {
		item := string(items1[i])
		items1Map[item] = true
	}

	// loop over items2 and store any duplicate items
	for i := 0; i < len(items2); i++ {
		item := string(items2[i])
		_, duplicate := items1Map[item]
		if duplicate {
			duplicateItemsMap[item] = true
		}
	}

	// create string of duplicate items
	duplicateItems := ""
	for item := range duplicateItemsMap {
		duplicateItems += item
	}

	return duplicateItems
}

func main() {
	// file with input rucksack contents
	filePath := "../rucksacks.txt"

	scanner, err := GetFileScanner(filePath)
	if err != nil {
		fmt.Println("failed to get file scanner", err.Error())
		return
	}

	sumOfPriorities := 0
	// scan file line-by-line
	for scanner.Scan() {
		rucksack := scanner.Text()
		halflen := utf8.RuneCountInString(rucksack) / 2
		// split rucksack items halfway to get compartment contents
		compartment1, compartment2 := rucksack[:halflen], rucksack[halflen:]

		// get duplicate items between two compartments
		duplicateItems := GetDuplicateItems(compartment1, compartment2)
		// return early if not exactly one duplicate is found
		duplicatesCount := utf8.RuneCountInString(duplicateItems)
		if duplicatesCount != 1 {
			fmt.Println(duplicatesCount, "duplicates found in rucksack", rucksack, "expected exactly 1")
			return
		}

		// update sum of item priorities
		sumOfPriorities += GetItemPriority(string(duplicateItems[0]))
	}

	fmt.Println(sumOfPriorities)

	scanner, err = GetFileScanner(filePath)
	if err != nil {
		fmt.Println("failed to get file scanner", err.Error())
		return
	}

	sumOfPriorities = 0
	// scan file n lines at a time
	n := 3
	for true {
		// slice of n rucksacks
		nRucksacks := make([]string, n)
		// condition for breaking out of outer loop
		breakOuterLoop := false

		// loop over and store n rucksack items
		for i := 0; i < n; i++ {
			// break out of inner and outer loops if no more lines are read
			if !scanner.Scan() {
				breakOuterLoop = true
				break
			}

			nRucksacks[i] = scanner.Text()
		}

		if breakOuterLoop {
			break
		}

		duplicateItems := nRucksacks[0]
		// loop over rucksack items and gather duplicate items
		for i := 1; i < n; i++ {
			duplicateItems = GetDuplicateItems(duplicateItems, nRucksacks[i])
		}

		// return early if not exactly one duplicate is found
		duplicatesCount := utf8.RuneCountInString(duplicateItems)
		if duplicatesCount != 1 {
			fmt.Println(duplicatesCount, "duplicates found in rucksacks", nRucksacks, "expected exactly 1")
			return
		}

		sumOfPriorities += GetItemPriority(string(duplicateItems[0]))
	}

	// update sum of item priorities
	fmt.Println(sumOfPriorities)
}

File: Day_3_Rucksack_Reorganization/Rust/main.rs
use std::fs;
use std::collections::HashSet;

// rucksack contents struct
struct Rucksack {
    // hash sets for rucksack compartment contents
    compartment1: HashSet<char>,
    compartment2: HashSet<char>,
}

impl Rucksack {
    // get duplicate items set between two compartments
    pub fn duplicate_items(&self) -> HashSet<char> {
        let duplicate_items_set: HashSet<char> = self.compartment1.intersection(&self.compartment2).cloned().collect();

        return duplicate_items_set;
    }

    // get set of all items in two compartments combined
    pub fn all_items(&self) -> HashSet<char> {
        let all_items_set: HashSet<char> = self.compartment1.union(&self.compartment2).cloned().collect();

        return all_items_set;
    }
}

// get priority of item based on ASCII value
fn get_item_priority(item: char) -> u8 {
    let item_ascii: u8 = item as u8;
    let lower_a_ascii: u8 = 'a' as u8;
    let upper_a_ascii: u8 = 'A' as u8;
    let lower_z_ascii: u8 = 'z' as u8;

    let item_priority: u8;
    // when item is lowercase, priority is given by how many steps ahead of 'a' the item is
    if item_ascii >= lower_a_ascii {
        item_priority = item_ascii - lower_a_ascii + 1;
    } else {
        // priority is given by how many steps ahead of 'A' the item is
		// plus the number of letters in the alphabet (i.e. 'z' - 'a' + 1)
        item_priority = item_ascii - upper_a_ascii + 1 + lower_z_ascii - lower_a_ascii + 1;
    }

    return item_priority;
}

fn main() {
    // file with input rucksack contents
    let file_path: &str = "../rucksacks.txt";
    // read file contents to string
    let file_contents: String = fs::read_to_string(file_path).expect("failed to read file");
    let mut rucksacks : Vec<Rucksack> = Vec::new();

    for line in file_contents.split("\n") {
        // get half length of line
        let half_len = line.len() / 2;
        // insert first half of line into compartment 1
        let mut compartment1: HashSet<char> = HashSet::new();
        for c in (&line[..half_len]).chars() {
            compartment1.insert(c);
        }
        // insert second half of line into compartment 2
        let mut compartment2: HashSet<char> = HashSet::new();
        for c in (&line[half_len..]).chars() {
            compartment2.insert(c);
        }

        // create rucksack with two compartments
        let rucksack: Rucksack = Rucksack {
            compartment1: compartment1,
            compartment2: compartment2,
        };

        // store ruckstack in vector
        rucksacks.push(rucksack);
    }

    let mut sum_of_priorities: u32 = 0;
    // iterate over rucksacks
    for rucksack in rucksacks.iter() {
        // get set of duplicate items between rucksack compartments
        let duplicate_items_set: HashSet<char> = rucksack.duplicate_items();
        // obtain first item in set as optional
        let first_item_optional: Option<&char> = duplicate_items_set.iter().next();
        match first_item_optional {
            // update sum of item priorities if first item is not None
            Some(item) => sum_of_priorities += get_item_priority(*item) as u32,
            // panic if first item is None
            None => panic!("failed to get common compartment items"),
        }
    }

    println!("{sum_of_priorities}");

    sum_of_priorities = 0;
    // iterate over rucksacks, n rucksacks at a time
    let n: usize = 3;
    for i in (0..rucksacks.len()).step_by(n) {
        // get set of all items in first rucksack's compartments
        // this is the overall set of duplicate items
        let mut duplicate_items_set: HashSet<char> = rucksacks[i].all_items();
        // iterate over next n - 1 rucksacks
        for rucksack in &rucksacks[(i + 1)..(i + n)] {
            // get set of all items in current rucksack's compartments
            let rucksack_all_items: HashSet<char> = rucksack.all_items();
            // retain only what is common between current rucksack and the overall set of duplicate items
            duplicate_items_set = duplicate_items_set.intersection(&rucksack_all_items).cloned().collect();
        }

        // obtain first item in set as optional
        let first_item_optional: Option<&char> = duplicate_items_set.iter().next();
        match first_item_optional {
            // update sum of item priorities if first item is not None
            Some(item) => sum_of_priorities += get_item_priority(*item) as u32,
            // panic if first item is None
            None => panic!("failed to get common compartment items"),
        }
    }

    println!("{sum_of_priorities}");
}